Рефакторинг после РК1

blackHATred opened the pull request.

blackHATred committed:
Рефакторинг;
Улучшена система ошибок;
Система ошибок вынесена в отдельный пакет;
Добавлены параметры в кофинг сервера;
Исправлена ошибка логаута;
Изменён валидатор Email;
Изменён эндпоинт для подборок по жанру;
Добавлен проброс контекста до ручек.

KristinaBu committed:
(1.6) тестовое покрытие 61%

blackHATred committed:
Create go.yml

blackHATred committed:
workflow

blackHATred committed:
workflow

blackHATred committed:
workflow

blackHATred committed:
workflow

blackHATred committed:
workflow

blackHATred committed:
coverage badge

blackHATred committed:
Изменён подход к хэшированию и хранения паролей;
Более продвинутые валидации пароля и емейла, соответствие валидации на фронте.

blackHATred committed:
parseParams.go перенесён в main.go;
Настройка docker;
Изменён роутинг, все эндпоинты теперь начинаются с /api/

blackHATred committed:
Исправлен тест

blackHATred committed:
Полный рефакторинг проекта после РК1

actions-user committed:
Update coverage badge

blackHATred committed:
Redis для хранения сессий;
Добавлено использование логгера echo.

blackHATred committed:
Merge pull request #4 from go-park-mail-ru/1.9-session-storage

1.9 session storage

KristinaBu committed:
тесты эндпоинты

KristinaBu committed:
Merge branch 'refactoring' of https://github.com/go-park-mail-ru/2024_1_Cyberkotletki into refactoring

blackHATred committed:
add new tests
add new error

blackHATred committed:
bugfix

ygimm commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+*.env
+go.sum
+config.yaml
```
предлагаю запушить конфиг (хотя бы для примера config.example.yaml)

blackHATred replied:
Там есть make для генерации примера конфига)

ygimm commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM golang:1.22.1-alpine3.19 AS build
```
респект за мультистейдж билд 

 хотелось бы увидеть сколько собирается контейнер и сколько он весит (пока у вас нет автодеплоя, не могу логи в ci посмотреть)

blackHATred replied:
Автодеплой пока в процессе, но контейнер с рабочим сервером (чисто скомпилированная гошка без редиса и nginx, которые находятся в отдельных контейнерах) весит ~10Мб

ygimm commented on internal/app/app.go:
```diff
@@ -2,72 +2,98 @@ package app
-	content.FilmsDatabase.InitDB()
+	})
+	// CORS
+	e.Use(middleware.CORSWithConfig(middleware.CORSConfig{
```
предлагаю добавить еще panic recovery и логирующую middleware. готовые брать нельзя)

ygimm commented on internal/delivery/http/auth.go:
```diff
@@ -0,0 +1,153 @@
+	key, err := h.useCase.Register(*registerData)
+	if err != nil {
+		switch {
+		// будем считать, что ErrAlreadyExists это тоже BadRequest
```
почему не 409 conflict?

blackHATred replied:
Как-то промахнулся с этим, забыл, что есть такой статус. Уже пофиксил

ygimm commented on internal/delivery/http/auth.go:
```diff
@@ -0,0 +1,153 @@
+	key, err := h.useCase.Register(*registerData)
+	if err != nil {
+		switch {
+		// будем считать, что ErrAlreadyExists это тоже BadRequest
```
почему не 409 conflict?

ygimm commented on internal/delivery/http/auth.go:
```diff
@@ -0,0 +1,153 @@
+	if err := decoder.Decode(registerData); err != nil {
+		return echoutil.NewError(c, http.StatusBadRequest, echoutil.BadJSON)
+	}
+	key, err := h.useCase.Register(*registerData)
```
на дженериках можно написать безоапсные утилитки для разыменования типа FromPtr
для перевода структуры в указатель тоже можно написать всякие Ptr, PtrOrNil и так далее

ygimm replied:
а почему бы не поменять интерфейс usecae просто?..

blackHATred replied:
Действительно, проще поменять интерфейс)
Так и сделал

ygimm commented on internal/delivery/http/auth_test.go:
```diff
@@ -0,0 +1,356 @@
+			// проверка
+			if tc.ExpectedErr != nil {
+				require.Error(t, err)
+				require.Equal(t, tc.ExpectedErr(c), err)
```
есть ErrorIs для сравнения ошибок (и можно будет более гибко с ними работать)

ygimm commented on internal/delivery/http/auth_test.go:
```diff
@@ -0,0 +1,356 @@
+	}
+}
+
+func TestAuthEndpoints_IsAuth(t *testing.T) {
```
а давйте распараллелим тесты?
не зря же мы табличное тестирвоание и моки придумали.
придется создавать моки и тестируемые сущности внтури каждого теста, но это все равно будет быстрее благодаря параллельным запускам тестов

ygimm commented on internal/delivery/http/content.go:
```diff
@@ -0,0 +1,43 @@
+	}
+	contentPreview, err := h.useCase.GetContentPreviewCard(int(id))
+	if err != nil {
+		if entity.Contains(err, entity.ErrNotFound) {
+			return echoutil.NewError(c, http.StatusNotFound, err)
+		} else {
+			return echoutil.NewError(c, http.StatusInternalServerError, err)
+		}
+	}
```
`if entity.Contains(err, entity.ErrNotFound) {
			return echoutil.NewError(c, http.StatusNotFound, err)
		} 
		return echoutil.NewError(c, http.StatusInternalServerError, err)`
почему не так? давайте уменьшать вложенность, код будет читаемее и проще
на все подобные места я не буду обращать внимание, правьте по мере нахождения.

к слову, линтер должен такое обнаруживать

ygimm replied:
оставил коммент про линтер, обратите внимание, пожалуйста

ygimm commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,30 @@
+    - name: Update coverage badge
+      run: |
+        coverage=$(go tool cover -func test.coverage | tail -n 1 | awk '{print $3}')
+        echo $coverage
```
а линтера у нас и нет))

давайте добавим линтер golangcilint. это стандартный линтер, много где используется. также давайте настроим yaml-файл для его конфигурации

ygimm commented on internal/repository/user.go:
```diff
@@ -0,0 +1,12 @@
+)
+
+//go:generate mockgen -source=$GOFILE -destination=mocks/mock_user.go
+type User interface {
```
мне нравится расположение интерфейсов и имплементации, по мне так идеальный баланс в холиварной теме о том, где размещать объявления интерфейсов))

ygimm commented on internal/repository/content/film.go:
```diff
@@ -0,0 +1,10 @@
+	"github.com/go-park-mail-ru/2024_1_Cyberkotletki/internal/entity"
+)
+
+type Film interface {
```
интерфейс с фильмом только уехал немного. 
подскажите, имплементация будет лежать еще глубже по иерархии проекта? 
`repository/content/film/impl.go`? типа такого?

blackHATred replied:
Задумка изначально была в том, чтобы подробить интерфейс Content на разные куски по типу Film и Series, которые должны лежать в папке content. От этой идеи сейчас отказался

ygimm reviewed (changes requested):
Очень хороший рефакторинг, очень хорошо написаны тесты (структура, нейминги и тд), реально кайфанул, пока читал код, хоть и есть парочка недостатков

есть два замечания, которые несложно поправить: 
- тесты не распараллелены
- нет линтера (и вытекающие отсюда стилистические недостатки)

линтер может увидеть больше стилистических ошибок, чем я (я по факту одну нашел), поэтому подключаем обязательно в ci (чтобы джоба падала при упавшем линтере) и кофнигурируем специальным файликом, чтобы он ругался в том числе и на отсутсвие комментов и на несортированные импорты и так далее

blackHATred committed:
Фиксы и улучшения:
Добавлен линтер;
Код отрефакторен в соответствии с линтером;
Изменены некоторые интерфейсы для взаимодействия через указатели;
Добавлен make для линтера;
Тесты распараллелены;
Добавлен middleware на recovery;
Добавлен middleware на requestID;
Улучшен middleware на CORS (переход со встроенного в echo на свой);
Интерфейс Film перенесён и расположен рядом с интерфейсом Content;
Более корректные тесты (gomock.Any() -> gomock.Cond())

blackHATred committed:
фикс линтера в github actions

blackHATred committed:
фикс t.Parallel();
добавлен недостающий эндпоинт GetCompilationByGenre.
